### PR TITLE
Refactor package and control API utility methods

### PR DESCRIPTION
**Pull request by sudarsanm94**
Refactor package and control API utility methods to make them easily configurable

### CODE REVIEW

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        if (upgradeRpm != 0) {
+            throw new PrestoManagerException("Failed to upgrade Presto", upgradeRpm);
+        }
+    }
+
+    public void startAsync()
+            throws PrestoManagerException
+    {
+        int startPresto = executeCommand(SUBPROCESS_TIMEOUT, "service", "presto", "start");
+        if (startPresto != 0) {
```
Comment (ShanBai6) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java]: Inline these

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+            LOGGER.error("Failed to start presto");
+        }
+    }
+
+    public void terminatePresto()
+            throws PrestoManagerException
+    {
+        int prestoTerminate = executeCommand("service", "presto", "stop");
+        if (prestoTerminate != 0) {
```
Comment (ShanBai6) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java]: ditto

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            throw new PrestoManagerException("Failed to stop presto", prestoTerminate);
+        }
+    }
+
+    public void killPresto()
+            throws PrestoManagerException
+    {
+        executeCommand("sudo", LAUNCHER_SCRIPT.toString(), "kill");
+        int prestoKill = executeCommand("service", "presto", "status");
+        if (prestoKill != 3) {
```
Comment (ShanBai6) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java]: ditto

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            throw new PrestoManagerException("Failed to stop presto", prestoKill);
+        }
+    }
+
+    public void restartAsync()
+            throws PrestoManagerException
+    {
+        int prestoRestart = executeCommand(SUBPROCESS_TIMEOUT, "service", "presto", "restart");
+        if (prestoRestart != 0) {
```
Comment (ShanBai6) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java]: ditto

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java [hunk 5]
```diff
@@ -17,78 +17,147 @@
...
     }
 
-    public Response startUsingRpm()
+    public Response uninstall(boolean checkDependencies, boolean forceUninstall)
     {
         try {
             if (!isInstalled()) {
                 LOGGER.error("Presto is not installed");
-                return Response.status(NOT_FOUND).entity("Presto is not installed").build();
+                return Response.status(BAD_REQUEST).entity("Presto is not installed").build();
```
Comment (jirassimok) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java]: `NOT_FOUND` was the correct status code here.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java [hunk 6]
```diff
@@ -201,88 +289,45 @@ public Response statusUsingRpm()
...
-    }
+    abstract void startAsync()
+            throws PrestoManagerException;
 
-    private static String getPrestoPort()
-            throws IOException
-    {
-        return getFileProperty(PRESTO_CONFIG_FILE, "http-server.http.port");
-    }
+    abstract void terminatePresto()
```
Comment (jirassimok) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java]: `terminatePresto`, `killPresto`, and `gracefulStop` should have more similar names.

I might just call them `terminate`, `kill`, and `stopGracefully` (or maybe`stop`).

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java [hunk 7]
```diff
@@ -17,78 +17,147 @@
...
 import static javax.ws.rs.core.Response.Status.ACCEPTED;
+import static javax.ws.rs.core.Response.Status.BAD_REQUEST;
 import static javax.ws.rs.core.Response.Status.CONFLICT;
 import static javax.ws.rs.core.Response.Status.INTERNAL_SERVER_ERROR;
 import static javax.ws.rs.core.Response.Status.NOT_FOUND;
 import static javax.ws.rs.core.Response.Status.OK;
 import static javax.ws.rs.core.UriBuilder.fromUri;
 
-public class PrestoRpmController
+public abstract class InstallationManager
```
Comment (jirassimok) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java]: I would split this class into an abstract class and one or two interfaces.

This class would still contain the methods that install, uninstall, and upgrade presto, and `getVersion`, but the other functionality it has would be delegated to the other class.

The interface (maybe `PrestoController`) would have the methods that start and stop Presto, and `isRunning`.

`RpmInstallationManager` could implement the interface, so it would not need any changes.

The only issue that arises is that the start/stop classes would not have access to `isInstalled`. To solve this, I would have the caller of the functions perform the check, which also helps by eliminating the repeated `if (!isInstalled) return NOT_FOUND` blocks in each of the start/stop functions.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java [hunk 8]
```diff
@@ -71,10 +71,8 @@ public static void replaceFile(Path path, String url)
             throws IOException
```
Comment (jirassimok) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java]: I would rename this method to `downloadFile`, and make its second argument a `URI`; let the caller create it however they like, not just as a string.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java [hunk 9]
```diff
@@ -130,4 +128,14 @@ public static void removePropertyFromFile(Path path, String property)
             properties.store(outputStream, null);
         }
     }
+
+    public static File downloadFile(URL url)
+            throws IOException
```
Comment (jirassimok) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java]: This should take a URI, not a URL (though in the function, you'll want to convert it to a URL with `URI#toURL()`). 

This could be named `downloadTempFile`, but `downloadFile` is about as good.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java [hunk 10]
```diff
@@ -130,4 +128,14 @@ public static void removePropertyFromFile(Path path, String property)
...
+    {
+        File tempFile = createTempFile("PrestoMgrTemp", ".tmp");
```
Comment (jirassimok) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java]: After importing `Files::createTempFile`:
```java
Path tempFile = createTempFile("PrestoMgrTemp", ".tmp");
```

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java [hunk 11]
```diff
@@ -36,6 +34,8 @@
 import java.util.stream.Stream;
 
 import static com.google.common.base.Preconditions.checkArgument;
+import static java.io.File.createTempFile;
```
Comment (jirassimok) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java]: Use `java.nio.file.Files.createTempFile` instead.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java [hunk 12]
```diff
@@ -36,6 +34,8 @@
...
+import static java.nio.file.Files.copy;
 import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
 import static java.util.stream.Collectors.collectingAndThen;
```
Comment (jirassimok) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java]: It's not functionally part of this pull request, but it'd be nice if the use of `collectingAndThen` were replaced by `ImmutableList.toImmutableList()`.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java [hunk 13]
```diff
@@ -17,78 +17,147 @@
...
 {
-    private static final Logger LOGGER = Logger.get(PrestoRpmController.class);
-    private static final String LAUNCHER_SCRIPT = "/usr/lib/presto/bin/launcher";
-    private static final Path PRESTO_CONFIG_FILE = Paths.get("/etc/presto/config.properties");
+    private static final Logger LOGGER = Logger.get(InstallationManager.class);
     private static final Gson GSON = new Gson();
```
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java]: This would be done while adding configuration

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java (hunk 3 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java]: I add the response code to PrestoManagerException.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java (hunk 8 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java]: This is not part of this pr

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java (hunk 9 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java]: This should be in a separate PR

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java (hunk 10 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java]: Done

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+    private static final Logger LOGGER = Logger.get(RpmInstallationManager.class);
+
+    public void installAsync(URL packageUrl, boolean checkDependencies)
+            throws PrestoManagerException
+    {
+        File tempFile = getRpmPackage(packageUrl);
+        try {
+            int installRpm;
+            if (checkDependencies) {
+                installRpm = executeCommand(SUBPROCESS_TIMEOUT, "sudo", "rpm", "-iv", tempFile.toString());
```
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java]: This can be modified in the future. I don't think this would reduce any code lines though.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java (hunk 5 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java]: Changed

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java [hunk 15]
```diff
@@ -17,78 +17,147 @@
...
-            try {
-                int startPresto = executeCommand(SUBPROCESS_TIMEOUT, "service", "presto", "start");
-                if (startPresto != 0) {
-                    LOGGER.error("Failed to start presto");
+        PrestoAsyncCommand(() -> uninstallAsync(checkDependencies));
+        return Response.status(ACCEPTED).entity("Presto is being uninstalled.\r\n" +
+                "To verify that uninstallation succeeded, check back later using the status API.").build();
+    }
+
+    public Response upgrade(String packageUrl, boolean checkDependencies, boolean preserveConfig, boolean forceUpgrade)
```
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java]: For upgrade presto not installed is not a problem. `rpm -U` doesn't throw any error if its already installed.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java [hunk 16]
```diff
@@ -107,7 +176,7 @@ public Response stopUsingRpm(StopType stopType)
                     killPresto();
                     break;
                 case GRACEFUL:
-                    if (isCoordinator()) {
+                    if (isRunningCoordinator()) {
                         LOGGER.error("Coordinator can't be gracefully stopped.");
                         return Response.status(CONFLICT).entity("Coordinator can't be gracefully stopped").build();
                     }
```
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java]: Then I can't send the appropriate response code.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java [hunk 17]
```diff
@@ -122,20 +191,40 @@ public Response stopUsingRpm(StopType stopType)
...
+        }
         return Response.status(OK).entity("Presto successfully stopped").build();
     }
 
-    public Response restartUsingRpm()
+    private void gracefulStop()
+            throws PrestoManagerException
+    {
+        try {
+            UriBuilder uriBuilder = fromUri(format("http://localhost:%s", getPrestoPort())).path("/v1/info/state");
```
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java]: What's the difference?

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java [hunk 18]
```diff
@@ -122,20 +191,40 @@ public Response stopUsingRpm(StopType stopType)
...
+        }
+        catch (ProcessingException e) {
+            LOGGER.warn(e, "Presto is not running");
+        }
+        catch (IOException e) {
+            throw new PrestoManagerException("Failed to get Presto port number.", e);
+        }
+    }
+
+    public Response restart()
```
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/InstallationManager.java]: I realized that `restart` should also be asynchronous since it starts the Presto again.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java (hunk 14 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmInstallationManager.java]: This should be done in the future PR

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageManager.java [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+    protected static final Path CATALOG_DIR = Paths.get("/etc/presto/catalog");
+    protected static final Path CONFIG_DIR = Paths.get("/etc/presto");
+    protected static final Path LAUNCHER_SCRIPT = Paths.get("/usr/lib/presto/bin/launcher");
+    protected static final int SUBPROCESS_TIMEOUT = 150;
+
+    public Response install(String packageUrl, boolean checkDependencies)
+    {
+        if ("".equals(packageUrl)) {
+            LOGGER.error("Url is empty or null");
+            return javax.ws.rs.core.Response.status(BAD_REQUEST).entity("Expected URL in the request body").build();
```
Comment (ShanBai6) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageManager.java]: Import the Response.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+
+    public static int getPrestoPort()
+            throws IOException
+    {
+        return Integer.valueOf(getFileProperty(PRESTO_CONFIG_FILE, "http-server.http.port"));
+    }
+
+    public static String getDiscoveryUri()
+    {
+        return null;
```
Comment (ShanBai6) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: Why is it returning null?

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmManager.java [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+import static com.teradata.prestomanager.agent.CommandExecutor.execCommandResult;
+import static com.teradata.prestomanager.agent.CommandExecutor.executeCommand;
+import static com.teradata.prestomanager.agent.PrestoConfigUtils.addTpchCatalog;
+import static com.teradata.prestomanager.agent.PrestoConfigUtils.deployConfigFiles;
+import static com.teradata.prestomanager.agent.PrestoConfigUtils.storeConfigFiles;
+import static java.lang.String.format;
+
+public class RpmManager
+        extends PackageManager
+        implements PrestoController
```
Comment (ShanBai6) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmManager.java]: This looks fancy. :+1:

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmManager.java [hunk 22]
```diff
@@ -0,0 +1,212 @@
...
+        if (upgradeRpm != 0) {
+            throw new PrestoManagerException("Failed to upgrade Presto", upgradeRpm);
+        }
+    }
+
+    public void startAsync()
+            throws PrestoManagerException
+    {
+        int startPresto = executeCommand(SUBPROCESS_TIMEOUT, "service", "presto", "start");
+        if (startPresto != 0) {
```
Comment (ShanBai6) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmManager.java]: What is the reason for not inline?

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/api/ControlAPI.java [hunk 23]
```diff
@@ -39,13 +39,7 @@
...
 public class ControlAPI
 {
-    private PrestoRpmController controller;
-
-    @Inject
-    public ControlAPI(PrestoRpmController controller)
-    {
-        this.controller = controller;
-    }
+    private static final PrestoController CONTROLLER = new RpmManager();
```
Comment (ShanBai6) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/api/ControlAPI.java]: Name it rpmController at least

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java (hunk 20 above)
Comment (jirassimok) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: If the method is new to the PR, either remove it, or give it a proper implementation.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java (hunk 8 above)
Comment (jirassimok) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/AgentFileUtils.java]: All right.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/api/ControlAPI.java (hunk 23 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/api/ControlAPI.java]: The control API class shouldn't know how it is implemented. That's why the general name.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmManager.java (hunk 22 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmManager.java]: I add it to the error message of PrestoMangerException

Comment (ShanBai6) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/RpmManager.java]: I see

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+            throws IOException
+    {
+        return Integer.valueOf(getFileProperty(PRESTO_CONFIG_FILE, "http-server.http.port"));
+    }
+
+    /**
+     * @return whether the Presto node is configured as coordinator
+     * @throws IOException If Presto port number could not be retrieved
+     */
+    public static boolean isConfiguredCoordinator()
```
Comment (rschlussel) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: where is this used?

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+            throws IOException
+    {
+        return Boolean.parseBoolean(getFileProperty(PRESTO_CONFIG_FILE, "coordinator"));
+    }
+
+    /**
+     * @return whether Presto is running as a coordinator
+     * @throws IOException If Presto port number could not be retrieved
+     */
+    public static boolean isRunningCoordinator()
```
Comment (rschlussel) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: Where is this used?

Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: In `stop()` method

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java (hunk 24 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: It's not used now. But it is needed for the controller. The agent will announce this along with the discovery  service.

Comment (rschlussel) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: This should get added with the commit that needs it.  The general rule is don't add methods until they're needed.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java (hunk 25 above)
Comment (rschlussel) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: the other method can be used instead, actually.  then you only need one method to check if something is a coordinator.  and it can be called isCoordinator().

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java (hunk 24 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: I'll remove this method

Comment (rschlussel) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: although actually i think you should keep this in place of isRunningCoordinator and call it isCoordinator (See comment below).

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java (hunk 25 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: This one is more accurate.
What if the user changes the config files and failed to restart Presto?

Comment (rschlussel) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: ok, so just get rid of the other method for now

Comment (jirassimok) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: The other method is going to be needed in an upcoming pull request.

Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoUtils.java]: I think you can add this method and `getDiscovery-uri()` method in an upcoming pull request
